Refine documentation for toolchain resolver plugins (#22132)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/toolchain_plugins.adoc

// Copyright 2022 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[toolchain_plugins]]
= Toolchain Resolver Plugins

Starting with Gradle version 7.6, JVM auto-provisioning logic gets decoupled from Gradle's lifecycle and moves into independent plugins, hosted on the https://plugins.gradle.org[Plugin Portal].

> For details on how toolchain auto-provisioning is using these plugins, see <<toolchains.adoc#sub:download_repositories,here>>.

Author: Good point -- I'll bring this link back in a similar form.

> In Gradle version 7.6 and above, Gradle provides a way to define toolchains in plugins.

Author: ```suggestion
In Gradle version 7.6 and above, Gradle provides a way to define Java toolchain auto-provisioning logic in plugins.
```

> This page explains how to author a toolchain resolver plugin.

Author: ```suggestion
This page explains how to author a toolchain resolver plugin.
For details on how toolchain auto-provisioning interacts with these plugins, see <<toolchains.adoc#sub:download_repositories,Toolchains>>.
```

== Provide a download URI

Toolchain resolver plugins provide logic to map a toolchain request to a download URI.

> They do this with an implementation of link:{javadocPath}/org/gradle/jvm/toolchain/JavaToolchainResolver.html[JavaToolchainResolver]:

Author: ```suggestion
They do this via an implementation of link:{javadocPath}/org/gradle/jvm/toolchain/JavaToolchainResolver.html[JavaToolchainResolver]:
```

        settings.getPlugins().apply("jvm-toolchain-management"); // <3>

        JavaToolchainResolverRegistry registry = getToolchainResolverRegistry();
        registry.register(JavaToolchainResolverImplementation.class); // <4>
        registry.register(JavaToolchainResolverImplementation.class);
    }
}
----
<1> The plugin needs to be a settings plugin and is abstract because it uses <<custom_gradle_types.adoc#property_injection,property injection>>.
<2> Access is needed to the link:{javadocPath}/org/gradle/jvm/toolchain/JavaToolchainResolverRegistry.html[JavaToolchainResolverRegistry] Gradle service to register the resolver implementation.
<3> These plugins are required to auto apply the `jvm-toolchain-management` base plugin, which dynamically adds the `jvm` block to `toolchainManagement`, thus making the registered toolchain repositories usable from the build.

> <4> The last step is to register the toolchain resolver with the registry service injected by Gradle.

Author: Thanks for catching this -- I thought I had removed it, but I must have forgotten to add it to the commit.
